### PR TITLE
Add workflows for screenbuilder commands

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
Using the **command-help.json** file located inside the resources folder of any CLI client a tip can be shown after specific commands to help guide the user into successfully creating a project

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:heart:

File: declarations.d.ts [hunk 1]
```diff
@@ -468,3 +468,40 @@ interface IBinaryPlistParser {
...
+	 * @return {IFuture<any>}      Object based on the JSON contents of the resource file.
+	 */
+	readJson(path: string): IFuture<any>;
+	/**
+	 * Builds the absolute path to a Cordova javascript file.
+	 * @param  {string} version  The Cordova version
+	 * @param  {string} platform The Platform - Android, iOS or WP8
+	 * @return {string}          Absolute path to Cordova javascript file
+	 */
+	buildCordovaJsFilePath(version: string, platform: string): string;
```
Comment (Fatme) [inline declarations.d.ts]: This function shouldn't be in common lib. It is specific only to ab-cli

File: services/commands-service.ts [hunk 2]
```diff
@@ -62,6 +73,11 @@ export class CommandsService implements ICommandsService {
 				} else {
 					command.execute(commandArguments).wait();
 				}
+				
+				if (this.commandHelp[commandName]) {
```
Comment (Fatme) [inline services/commands-service.ts]: `commandHelp` function returns the content of `this.$staticConfig.COMMAND_HELP_FILE_NAME` but here we use it as dictionary.. I'm confused :confused:

Comment (Mitko-Kerezov) [inline services/commands-service.ts]: the contents of `this.$staticConfig.COMMAND_HELP_FILE_NAME` are read as JSON and `commandHelp` is defined as a `get` property so it is basically a dictionary

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: resource-loader.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+///<reference path=".d.ts"/>
+"use strict";
+
+import path = require("path");
```
Comment (teobugslayer) [inline resource-loader.ts]: consider using the ES6 syntax here

File: resource-loader.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+export class ResourceLoader implements IResourceLoader {
+	constructor(private $fs: IFileSystem,
+		private $staticConfig: Config.IStaticConfig) { }
+
+	resolvePath(resourcePath: string): string {
+		return path.join(__dirname, "../../resources", resourcePath);
+	}
+
+	openFile(resourcePath: string): any {
```
Comment (teobugslayer) [inline resource-loader.ts]: This function should return a ReadStream. You may also fix the code in file-system.ts

File: services/commands-service.ts [hunk 5]
```diff
@@ -19,14 +19,27 @@ export class CommandsService implements ICommandsService {
 	private static HIERARCHICAL_COMMANDS_DEFAULT_COMMAND_DELIMITER = "|*";
 	private static HOOKS_COMMANDS_DELIMITER = "-";
 	private areDynamicSubcommandsRegistered = false;
+	private cachedCommandHelp: any = {};
```
Comment (teobugslayer) [inline services/commands-service.ts]: Can we use undefined or null instead of {} for the uninitialized value?

File: services/commands-service.ts [hunk 6]
```diff
@@ -19,14 +19,27 @@ export class CommandsService implements ICommandsService {
...
-		private $commandsServiceProvider: ICommandsServiceProvider,
-		private $options: ICommonOptions) {
+		private $injector: IInjector,
+		private $logger: ILogger,
+		private $options: ICommonOptions,
+		private $resources: IResourceLoader,
+		private $staticConfig: Config.IStaticConfig) {
+	}
+
+	public getCommandHelp(): IFuture<any> {
```
Comment (teobugslayer) [inline services/commands-service.ts]: Can we add unit tests?

File: resource-loader.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+		return this.$fs.createReadStream(this.resolvePath(resourcePath));
+	}
+
+	readJson(resourcePath: string): IFuture<any> {
+		return this.$fs.readJson(this.resolvePath(resourcePath));
+	}
+
+	public buildCordovaJsFilePath(version: string, platform: string): string {
```
Comment (ligaz) [inline resource-loader.ts]: Isn't this specific to AppBuilder CLI only?

File: services/commands-service.ts [hunk 8]
```diff
@@ -19,14 +19,27 @@ export class CommandsService implements ICommandsService {
...
+		return (() => {
+			if (_.isEmpty(this.cachedCommandHelp) && this.$fs.exists(this.$resources.resolvePath(this.$staticConfig.COMMAND_HELP_FILE_NAME)).wait()) {
```
Comment (ligaz) [inline services/commands-service.ts]: Actually this is a very common pattern - lazy instantiation. Consider creating a `@lazy` decorator in order to abstract this code.

**Comment by justcodebuilduser** [comment]
:heart:

File: services/commands-service.ts [hunk 9]
```diff
@@ -295,6 +304,16 @@ export class CommandsService implements ICommandsService {
 		}).future<boolean>()();
 	}
 
+	private getCommandHelp(): IFuture<any> {
+		return (() => {
+			if (_.isEmpty(this.cachedCommandHelp) && this.$fs.exists(this.$resources.resolvePath(this.$staticConfig.COMMAND_HELP_FILE_NAME)).wait()) {
```
Comment (teobugslayer) [inline services/commands-service.ts]: this.cachedCommandHelp should be null, and the isEmpty check is redundant. You can simply check for existence.

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: resource-loader.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+"use strict";
+
+import * as path from "path";
+
+export class ResourceLoader implements IResourceLoader {
+	constructor(private $fs: IFileSystem,
+		private $staticConfig: Config.IStaticConfig) { }
+
+	resolvePath(resourcePath: string): string {
+		return path.join(__dirname, "../../resources", resourcePath);
```
Comment (rosen-vladimirov) [inline resource-loader.ts]: this points to a dir, that is not part of the common lib, which makes this code unusable when you want just the common-lib

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by teobugslayer** [comment]
:+1:

**Pull request merged**